annotated.go

```diff
@@ -89,3 +91,296 @@ func (a Annotated) String() string {
+
+// annotationError is a wrapper for an error that was encountered while
+// applying annotation to a function. It contains the specific error
+// that it encountered as well as the target interface that was attempted
+// to be annotated.
+type annotationError struct {
+	target interface{}
+	err    error
+}
+
+func (e *annotationError) Error() string {
+	return e.err.Error()
+}
+
+type paramTagsAnnotation struct {
+	tags []string
+}
+
+var _ paramTagsAnnotation = paramTagsAnnotation{}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.In
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (pt paramTagsAnnotation) apply(ann *annotations) error {
+	if ann.annotatedIn {
+		return fmt.Errorf("cannot apply more than one line of ParamTags")
+	}
+	ann.annotatedIn = true
+	fType := ann.fType
+
+	annotatedParams := []reflect.StructField{{
+		Name:      "In",
+		Type:      reflect.TypeOf(In{}),
+		Anonymous: true,
+	}}
+	for i := 0; i < fType.NumIn(); i++ {
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.In(i),
+		}
+		if i < len(pt.tags) {
+			structField.Tag = reflect.StructTag(pt.tags[i])
+		}
+		annotatedParams = append(annotatedParams, structField)
+	}
+
+	ann.Ins = []reflect.Type{reflect.StructOf(annotatedParams)}
+	return nil
+}
+
+// ParamTags is an Annotation that annotates the parameter(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional parameter.
+func ParamTags(tags ...string) Annotation {
+	return paramTagsAnnotation{tags}
+}
+
+type resultTagsAnnotation struct {
+	tags []string
+}
+
+var _ resultTagsAnnotation = resultTagsAnnotation{}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.Out
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (rt resultTagsAnnotation) apply(ann *annotations) error {
+	if ann.annotatedOut {
+		return fmt.Errorf("cannot apply more than one line of ResultTags")
+	}
+	ann.annotatedOut = true
+	fType := ann.fType
+
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	returnsError := false
+	for i := 0; i < fType.NumOut(); i++ {
+		if fType.Out(i) == _typeOfError {
+			returnsError = true
+			continue
+		}
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.Out(i),
+		}
+		if i < len(rt.tags) {
+			structField.Tag = reflect.StructTag(rt.tags[i])
+		}
+		annotatedResult = append(annotatedResult, structField)
+	}
+	if returnsError {
+		ann.Outs = []reflect.Type{
+			reflect.StructOf(annotatedResult),
+			_typeOfError,
+		}
+	} else {
+		ann.Outs = []reflect.Type{reflect.StructOf(annotatedResult)}
+	}
```

Comment:
optional simplification:

```suggestion
	ann.Outs = []reflect.Type{reflect.StructOf(annotatedResult)}
	if returnsError {
		ann.Outs = append(ann.Outs, _typeOfError)
	}
```